Cover all possible values for venstar operation_mode

flamechair opened the pull request:
## Description:
https://github.com/home-assistant/home-assistant/blob/2c00ff7e522ccb16a023816c313f4747d718810d/homeassistant/components/venstar/climate.py#L269

Means that operation mode can be one of either:
 `[HVAC_MODE_HEAT, HVAC_MODE_COOL, HVAC_MODE_OFF, HVAC_MODE_AUTO]` from `homeassistant.const`
or 
`self.MODE_OFF, self.MODE_HEAT, self.MODE_COOL, self.MODE_AUTO` (which evaluate to `0, 1, 2, 3` respectively from `venstarcolortouch`)

However in our conditional check, we only look for: `self.MODE_OFF, self.MODE_HEAT, self.MODE_COOL, self.MODE_AUTO`


https://github.com/home-assistant/home-assistant/blob/2c00ff7e522ccb16a023816c313f4747d718810d/homeassistant/components/venstar/climate.py#L277-L283

This PR ensures that we check for all possible options

**Related issue (if applicable):** None

## Example entry for `configuration.yaml` (if applicable):
```yaml
climate:
  - platform: venstar
    host: !secret venstar_host
    humidifier: true
    timeout: 300
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [ ] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [ ] Untested files have been added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

Commit:
cover all possible values for operation_mode

Commit:
Update climate.py

Commit:
Update climate.py

MartinHjelmare commented on homeassistant/components/venstar/climate.py:
```diff
@@ -275,11 +275,11 @@ def set_temperature(self, **kwargs):
 
         if set_temp:
-            if operation_mode == self._client.MODE_HEAT:
+            if operation_mode in (self._client.MODE_HEAT, HVAC_MODE_HEAT):
```
Operation mode should only be a home assistant specific constant, not a constant on the client.

Please change that.

Commit:
Update climate.py

mapped homeassistant constants to client ones so we don't have to check for both

Commit:
black + pylint

MartinHjelmare commented on homeassistant/components/venstar/climate.py:
```diff
@@ -266,20 +266,25 @@ def _set_operation_mode(self, operation_mode):
         temp_low = kwargs.get(ATTR_TARGET_TEMP_LOW)
         temp_high = kwargs.get(ATTR_TARGET_TEMP_HIGH)
         temperature = kwargs.get(ATTR_TEMPERATURE)
+        mode_map = {
```
We could move this to init method since it shouldn't change.

MartinHjelmare reviewed (commented):
Looks good! Just a clean up possible.

Commit:
move mode_map to __init__()

MartinHjelmare reviewed (approved):
Great!

iamtpage commented:
Since we're modifying the default value of `operation_mode`, would this be considered a breaking change? If so, could you add the label since I'm unable to?

cc: @MartinHjelmare

MartinHjelmare commented on homeassistant/components/venstar/climate.py:
```diff
@@ -266,20 +271,20 @@ def _set_operation_mode(self, operation_mode):
 
-        if operation_mode != self._client.mode:
-            set_temp = self._set_operation_mode(operation_mode)
+        if self._mode_map.get(operation_mode) != self._client.mode:
```
```suggestion
        if operation_mode and self._mode_map.get(operation_mode) != self._client.mode:
```

MartinHjelmare commented on homeassistant/components/venstar/climate.py:
```diff
@@ -266,20 +271,20 @@ def _set_operation_mode(self, operation_mode):
 
         if set_temp:
-            if operation_mode == self._client.MODE_HEAT:
+            if operation_mode == HVAC_MODE_HEAT:
```
Here we should check against the existing client mode instead, ie either the new mode that we just set, or if we didn't change mode, the existing client mode.

iamtpage replied:
Why? I'm not sure I understand the reasoning here

MartinHjelmare replied:
We want to set the temp set points according to the current client mode of the device, right? If the user didn't pass an operation mode argument we will not set the correct set point with the current code. We need to change to check the mode of the device instead.

iamtpage replied:
Ah got it. So instead of comparing `operation_mode` to `HVAC_MODE_HEAT`, we'd compare `self._client.mode` to `self._client.MODE_HEAT` right?

MartinHjelmare replied:
Yes! Has `self._client.mode` had time to update correctly if we asked the libraryt to change it in the call above?

iamtpage replied:
No, if `operation_mode` is `None` then it won't ever pass the conditional check to change the operation mode.

 If `operation_mode` is not `None` and `set_temp` is `True` after we try to change the operation mode, then we can safely say that `operation_mode == self._mode_map.get(operation_mode)`

If `operation_mode` is `None` we should have no reason to believe that our value of `self._client.mode` is out of date because we haven't done anything with it. So we could safely compare `self._client.mode` to `self._client.(MODE_HEAT|MODE_COOL|MODE_AUTO)`

Or if we wanted to ensure we had the most up to date value for `self._client.mode` we'd have to either wait for another `update()` cycle or manually call `self._client.update_info()` before the comparison.

The 3rd option seems best, but runs the risk of having old data since the values are cached locally until you call `self._client.update_info()` again

MartinHjelmare replied:
Can we check against `self._mode_map.get(operation_mode)` if that is not None or `self._client.mode`?

That seems simplest.

MartinHjelmare reviewed (changes requested):
Sorry, I read too fast before.

Commit:
Update climate.py

Commit:
Update climate.py

MartinHjelmare reviewed (approved):
Nice!

MartinHjelmare commented:
We're good to merge now, right?

iamtpage commented:
Should be. I had to step away before lining finished. Once it passes black + pylint we will be good

-------- Original Message --------
On Oct 23, 2019, 6:56 PM, Martin Hjelmare wrote:

> We're good to merge now, right?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, [view it on GitHub](https://github.com/home-assistant/home-assistant/pull/27949?email_source=notifications&email_token=ACD5GLU7MPZ6SJOITV7SWB3QQDQDLA5CNFSM4JCSCBHKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOECDHTRA#issuecomment-545683908), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/ACD5GLU5VVL4BXEHCPRKUHLQQDQDLANCNFSM4JCSCBHA).

iamtpage commented:
G2G, feel free to merge whenever @MartinHjelmare
